### ISSUE TITLE
[BUG] Restoring snapshot: indices exclusion triggers security_exception (creating OK, listing OK)

### ISSUE THREAD

**Issue by ict-one-nl** [opensearch-project/security#1652]
**Describe the bug**
I'm using the S3 repository plugin to store snapshots. I've have tested this previously with the same scripting, but now I can't restore snapshots anymore. Don't know the exact cause, two things have changed: I have moved to 1.2.4 and I have moved from SAML to openid. The calls for the snapshot create/restore/list/etc are still being done through basic auth. 

The strange thing is, I can list all the snapshots, I can create snapshots, I just can't restore them. I'm not including global state or the security index:

Error:

```
{
    "error": {
        "root_cause": [
            {
                "type": "security_exception",
                "reason": "no permissions for [] and User [name=admin_XXXX, backend_roles=[admin], requestedTenant=null]"
            }
        ],
        "type": "security_exception",
        "reason": "no permissions for [] and User [name=admin_XXXX, backend_roles=[admin], requestedTenant=null]"
    },
    "status": 403
}
```

Listing works:

![image](https://user-images.githubusercontent.com/2643715/151875592-0ac3f279-9e2b-4479-b5b5-d4215af8eb39.png)

Creating snapshot works:

![image](https://user-images.githubusercontent.com/2643715/151875665-2bbbb8dc-dcaf-4e54-bc50-7c8b36ce93c3.png)

Restoring fails:

![image](https://user-images.githubusercontent.com/2643715/151876467-20c2fbdb-fc27-4387-8637-c6da2c927b9d.png)

all_access is mapped to the admin backend role:

![image](https://user-images.githubusercontent.com/2643715/151875796-df14f3ab-adb0-4379-9c44-2b8803b8d739.png)

**Expected behavior**
Snapshot is restored

**Plugins**
Default docker 1.2.4 plus s3 repo plugin
Also tried default docker 1.1 plus s3 repo plugin

**Screenshots**

**Host/Environment (please complete the following information):**
Docker 1.2.4 image on kubernetes

**Additional context**

**Comment by ict-one-nl** [comment]
Would be nice if a bug fix could make it to the next release. For now I just snapshot the stuff that is really really necessary, but it would be nice to just do it the default way.

**Comment by JeffBolle** [comment]
This appears to impact 1.3 as well.

Even adding the snapshot_restore and manage_snapshot roles to our admin users via OpenID did not resolve the issue. I was able to get past it by using the admin certificate.

**Comment by peternied** [comment]
@cliu123 Could you look into this issue and recommend next steps?

**Comment by igorneos** [comment]
Any news?

**Comment by peternied** [comment]
@cliu123 did you have a chance to look into this?

**Comment by ict-one-nl** [comment]
Guys, with 2.0.1 my workaround with restoring specific indexes doesn't work anymore. Basically I can't recover my cluster with my default admin credentials.

I do not want to be pushy, but this is getting **really** annoying and breaks the snapshotting functionality completely......

![image](https://user-images.githubusercontent.com/2643715/175016208-9add7316-05d2-41f3-84cd-232d9cf613ac.png)

![image](https://user-images.githubusercontent.com/2643715/175016522-639c6f6e-e4b4-490f-b78c-fc117f8ba52e.png)

**Comment by ict-one-nl** [comment]
Trick with the admin certificate works though.

**Comment by ict-one-nl** [comment]
Guys... although we have solved it by using certificates this is quite an irritating bug, if you don't understand what's happening you won't be able to restore. When people are already in panic with a cluster that has exploded this can be frustrating.

**Comment by smlx** [comment]
@ict-one-nl could you elaborate exactly what the workaround is for this?

**Comment by ict-one-nl** [comment]
Use certificate based authentication instead of basic auth, then it won't be a problem. We switched over to certificate based auth because of this.

**Comment by hquan5397** [comment]
@ict-one-nl Can you provide how to generate certificate to perform certificate based authentication in AWS

**Comment by ict-one-nl** [comment]
We have deployed it locally in kubernetes, so I'm afraid I have little knowledge of the AWS setup :)

**Comment by ict-one-nl** [comment]
Guys, any news on this? Although there is a workaround this will give more novice users a headache.

**Comment by hquan5397** [comment]
@ict-one-nl I set "include global state" to false then i can take a snapshot

**Comment by ict-one-nl** [comment]
Problem is restoring it :) See the first posts, all requests to validate this are present :)

**Comment by hquan5397** [comment]
> Problem is restoring it :) See the first posts, all requests to validate this are present :)

Oh that's my mistake. i mean i can restore it if i set include global state to false @ict-one-nl

**Comment by scrawfor99** [comment]
[TRIAGE] @scrawfor99 follow-up with poster.

**Comment by scrawfor99** [comment]
HI @ict-one-nl, I wanted to reach out to share this documentation link with you and see if this was at all helpful. Please feel free to follow-up if you have further questions or concerns. 

https://opensearch.org/docs/1.2/opensearch/snapshot-restore/#security-plugin-considerations

**Comment by ict-one-nl** [comment]
Yes I'm aware of the limitations, global state and the opendistro security index are mentioned in the bug report already. To what are you referring specifically?

**Comment by b-lancaster** [comment]
Any update on this? We're facing the same problem.

**Comment by bdbrink** [comment]
Seeing the same issue with version 1.3, tried the work arounds but the same permission errors persist.

**Comment by b-lancaster** [comment]
After seeking support from AWS, we were told to also exclude `.kibana*` from the indices as well and that fixed our problem. Hope this helps someone else.

```
POST /_snapshot/<repository-name>/<snapshot-name>/_restore
{
  "indices": "-.kibana*,-.opendistro*"
```

**Comment by b-lancaster** [comment]
@ict-one-nl Good point. I didnt think about how our setup may differs from others. All of our saved objects like visualizations etc are deployed to our cluster via terraform. So we were able to just redeploy all of that after we got the indices restored. 

Also, I'm no expert in OpenSearch. I just know that it worked lol.

**Comment by ict-one-nl** [comment]
Problem with that can be the upgrades, we imported suricata visualisations/dashboards into 1.x, worked fine. Restoring the same dashboards to 2.x through the dashboards api failed. Upgrading 1.x to 2.x would have saved them. 

We would have liked to be able to back these up instead of doing another migration path or recreating them. A well functioning backup/restore is a must have in my opinion, you need to be able to do disaster recovery fast.

**Comment by Ash-Garg** [comment]
I am facing the same issue , trying to move from AWS Managed elasticsearch to AWS Managed opensearch . however this issue is preventing the migration.

Below solution works but not ideal :
POST /_snapshot/<repository-name>/<snapshot-name>/_restore
{
  "indices": "-.kibana*,-.opendistro*,-.tasks*" }

**Comment by cwperks** [comment]
[Triage] @davidlago Would you please take a look at this issue?

**Comment by scrawfor99** [comment]
Update 12/13: I went ahead and setup a Kubernetes deployment to begin the diagnosis process. I am going to reproduce the issue and start moving out from the `snapshotRestoreEvaluator` as this seems central to the issue.

**Comment by ict-one-nl** [comment]
Sounds great, thanks for the effort :)

**Comment by scrawfor99** [comment]
Update 12/14:

1. Completed setup of deployment using minikube, helm, postman. Am in the process of creating demo snapshot to test restore.

**Comment by scrawfor99** [comment]
Update 12/15: 

I have reproduced the main errors listed in the top couple of posts of this issue. I have also verified that the configuration is valid. That being said, I ran into a couple of differences from the original issue:

- I did not need to exclude any indices to reproduce the error. It seems like the issue happens anytime you try to restore a snapshot regardless of inclusion/exclusion of indices. 
- I also found that deleting the kibana_1 index did not resolve the issue. I was still unable to restore my demo snapshot after deleting every index but the security index in fact.

**Comment by scrawfor99** [comment]
Update 12/16: I believe I have narrowed down the issue to the two most likely causes. As can be seen in the first attachment, when attempting to restore an index the ﻿﻿﻿SnapshotRestoreHelper is unable to read the information from the snapshot repository. In this case, I had 7 snapshots (the loop happened 7 times which is correct if the name is not found), but the names are not shown in the logs as they should be. The correct formatting of the log is also verified by the original code which reads `log.debug("snapshot found: {} (UUID: {})", snapshotId.getName(), snapshotId.getUUID());` -- this shows it is not a datatype issue causing the log to not display the info. 

Another possible issue is in the resolveIndexPatterns function of the IndexResolverReplacer. In the second attachment, you can see that the requested patterns has swapped from my original input of snapshot '3' to '*'. This would not be allowed by anyone other than a super admin (requiring the admin certificate) as we can see in the ProtectedIndexAccessEvalutor that requests for '_all' indices are not allowed by a regular user (this includes the admin user). 

I am not yet sure which of these two problems is the root issue or if it is a combo of the two but hope this provides some insight and that you know it is being addressed. 

![Uploading Screen Shot 2022-12-16 at 3.16.05 PM.png…]()
<img width="1338" alt="Screen Shot 2022-12-16 at 3 22 13 PM" src="https://user-images.githubusercontent.com/65832608/208182738-9c7d13c8-e493-4f0e-b6c7-3e0b1a9e880c.png">

**Comment by ict-one-nl** [comment]
Great work, good that other stuff became apparent as well. Next week I won't have access to a computer much, but if I can help after that week let me know.

**Comment by scrawfor99** [comment]
Update 12/19: I added further logs today and went through trying to determine the exact flow of a request and where we could be running into an issue that would cause the request information to be lost. From what I could find, the earliest step of the process which is RestoreSnapshot specific is the Snapshot Restore Evaluator at which point we see that the request is still null. Continuing forward, I found that throughout the entire flow of the request the request information evaluates to null. To the best of my understanding this should not be the case. It seems like the Thread name is not being set either.

Attached is another snippet of the logs from my testing. A relevant section to this particular issue with the security exception you are encountering is that after failing to find `snapshot repository 'my-fs-repository', snapshot 'my_snapshot_no_kib'` the IndexResolverReplacer interprets the request as looking for [*] which is a Local All pattern and would be blocked from non-super-admins. 

<img width="1500" alt="Screen Shot 2022-12-19 at 4 53 32 PM" src="https://user-images.githubusercontent.com/65832608/208531320-b0c3d8a8-8efd-4b8e-9c50-074c6ce66b24.png">

**Comment by scrawfor99** [comment]
Update 12/20: I went through the logs and found the point where the request gets swapped from its original form to the wildcard all. I believe it happens when the DlsFlsValveImpl invokes() the snapshot restore request. At this point the request elements read correctly as `"indices" : ["-.opendistro_security"]` but then after the resolved request becomes `aliases=[*], allIndices=[*]`. I could be mis-interpretting this but I will be looking into this specific exchange next.

**Comment by scrawfor99** [comment]
Update 12/21: Went through logs further and tried bypassing a check for marking complete the presponse in SnapshotRestoreEvaluator. This did not fix the issue but did allow me to see further steps of the execution path. I added further logs to the execution path to look at possible jumps in the execution over needed request logic. Will continue to work on the issue and may bring in another contributor.

**Comment by scrawfor99** [comment]
Update 12/22: @cwperks lent me a second pair of eyes and a different angle today and was able to determine that the issue appears to be coming from the parsing of the indices done by the opensearch-core snapshot utils. Basically when it is reading the indices, it treats the ordering differently based on whether a wildcard statement appears first or not in the `indices` field. Craig is going to add additional context to this later today and then request a transfer of the ticket. That being said, because this issue has been long standing, I am going to continue to prioritize a fix based on the core end. That way the issue does not get buried as it had before.

**Comment by cwperks** [comment]
TLDR; Order matters in the `indices` section of the request. The issue is coming from SnapshotUtils in core when a negated index is in the front. Relevant lines: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/snapshots/SnapshotUtils.java#L91-L95

To find the root cause of the issue I added a new debugging statement after this line to print out the resolved indices: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/SnapshotRestoreEvaluator.java#L97

Then I ran the following queries on a snapshot that contained:

```
.opendistro_security
.kibana_1
opensearch_dashboards_sample_data_flights
security-auditlog-2022.12.22
``` 

These were the requests and resolved indices:

```
// -.opendistro_security in the front of the list
POST _snapshot/my-first-repo/20221222-snapshot-all/_restore
{
    "indices": ["-.opendistro_security", "open*", "sec*"],
    "include_global_state" : false
}

// Restoring indices: [opensearch_dashboards_sample_data_flights, security-auditlog-2022.12.22, .kibana_1]
```

```
// -.opendistro_security in the middle of the list
POST _snapshot/my-first-repo/20221222-snapshot-all/_restore
{
    "indices": ["open*", "-.opendistro_security", "sec*"],
    "include_global_state" : false
}

// Restoring indices: [opensearch_dashboards_sample_data_flights, security-auditlog-2022.12.22]
```

Wait a minute, why do these resolve to a different set of indices???

That led us to look into SnapshotUtils which is resolving the request to a list of indices to resolve and that's where we found specific logic for when a negation was the first in the list (https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/snapshots/SnapshotUtils.java#L91-L95):

```
// if its the first, fill it with all the indices...
if (i == 0) {
    result = new HashSet<>(availableIndices);
}
```

All `availableIndices` are being added to the `HashSet` and the negated index is removed later on here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/snapshots/SnapshotUtils.java#L115

The thing is, if the negation is later on in the list then it does not add all available indices to the set and only removes the index in the negation if it was previously added to the HashSet.

The quickest solution I can think of is to add sorting logic in `SnapshotUtils.filterIndices` to sort `selectedIndices` so that negations always appear at the end of the array.

**Comment by cwperks** [comment]
The issue really is with a mix of positive (`my-index*`) and negative (`-.opendistro_security`) patterns in the `indices` portion.

If `indices` only contains negative. i.e. `"indices": ["-.opendistro_security", "-.kibana_1"]` then its interpreted as everything except these 2 indices.

If `indices` only contains positive. i.e. `"indices": ["open*",  "sec*"]` then its interpreted as everything matching these patterns.

It gets tricky with a mix of positive and negative. i.e. `"indices": ["-.opendistro_security", "open*"]` which is where this behavior arose. If you only want indices matching a certain pattern then you don't need the negation in the restore snapshot query.

I'd like to think through scenarios, but when there is both positive and negative then I think the positive should be evaluated first and the negative taken away from what those have resolved to.

**Comment by ict-one-nl** [comment]
Thanks again for all the effort guys, sounds like you're homing in on the actual problem. Great work!

**Comment by scrawfor99** [comment]
Opened a PR to resolve the issue: https://github.com/opensearch-project/OpenSearch/pull/5626/files

**Comment by scrawfor99** [comment]
@ict-one-nl, @davidlago , @cwperks , @CEHENKLE 

I am happy to share that the PR fixing this behavior has been merged into the main branch of the core repository and will be backported. I want to take the opportunity to thank you for your patience in the resolution of this issue since its original opening as well as taking the time to attend our public triaging meeting to bring further attention to this issue. 

Supporting our community is always the number one priority. I have copied over the updated functionality so that you know what to expect when making calls under the corrected restore logic. 

The new behavior is as follows: 

1. (Querying a single negated snapshot): If the body of the request reads `indices: "-bar"`, the query will return all indices which are not `"bar"`. 
2. (Querying a single negated snapshot wildcard): If the body of the request reads `indices: "-bar*"`, the query will return all indices which are not a subset of `"bar*"`. 
3. (Querying multiple negated snapshots): If the body of the request reads `indices: "-bar, -baz"`, the query will return all indices which are not `"bar"` or `"baz". 
4. (Querying multiple negated snapshot wildcards): If the body of the request reads `indices: "-bar*, -baz*"`, the query will return all indices which are not a subset of `"bar*"` or a subset of `"baz*". 
5. (Querying a single snapshot): If the body of the request reads `indices: "bar"`, the query will return `"bar"`. 
6. (Querying multiple snapshot wildcards): If the body of the request reads `indices: "bar*, baz*"`, the query will return all indices which are a subset of `"bar*"` or a subset of `"baz*". 
7. (Querying multiple mixed snapshot wildcards): If the body of the request reads `indices: "-bar*, ba*"`, the query will return all indices which are a subset of `"ba*"` but not a subset of `"bar*"`. **This was not the case previously.**
8. (Querying multiple mixed snapshot wildcards): If the body of the request reads `indices: "ba*, -bar*"`, the query will return all indices which are a subset of `"ba*"` but not a subset of `"bar*"`. 

Now behavior is persevered irregardless of the order of positive and negated wildcards queries.  

Thank you and Happy New Year!

**Comment by peternied** [comment]
@scrawfor99 Did we add a test that verifies the bug is fixed in the security repo?

**Comment by scrawfor99** [comment]
@peternied we have not. If that is something you think would be a good idea I can go ahead and add a test.

**Comment by peternied** [comment]
If we can I'd recommend we test in scenarios that most closely reflect what customers encounter, I'd hate for another change in core to go in that seems innocuous but it breaks snapshot restore via the security plugin

**Comment by ict-one-nl** [comment]
nice thank you!

**Comment by scrawfor99** [comment]
@peternied, I opened a PR in the security codebase to add a test which will check that the behavior is consistent during the SnapshotRestore operation. So once this gets merged everything should be set and this issue will be double covered with new testing in core and security.

**Comment by ict-one-nl** [comment]
Has the PR been merged? I just tried to do a disaster recovery as a test and encountered a very familiar error:

![image](https://user-images.githubusercontent.com/2643715/224166890-8505682a-c85f-4043-b010-c015fcfe743d.png)

![image](https://user-images.githubusercontent.com/2643715/224166966-c9f147dc-2e26-4935-b648-43d64646c26d.png)

:)

**Comment by peternied** [comment]
@scrawfor99 When you get a chance can you confirm if your change was in 2.6.0?

**Comment by cwperks** [comment]
@peternied @scrawfor99 I left a comment on the PR in core. The backport label was not added to the PR so it was not backported and released in 2.6.0. The soonest it will be released is 2.7.0.

**Comment by ict-one-nl** [comment]
It's on the list for 2.7? cool :)
